File: programmers/월간 코드 챌린지 시즌3/금과 은 운반하기/solution.js
function canBuild(a, b, g, s, w, t, T) {
  let [S, G, total] = [0, 0, 0]

  t.forEach((t, i) => {
    let W = Math.min(g[i] + s[i], w[i] * Math.round(T / (2 * t)))

    S += Math.min(s[i], W)
    G += Math.min(g[i], W)
    total += W
  })

  return a <= G && b <= S && a + b <= total
}

function solution(a, b, g, s, w, t) {
  let left = 1
  let right = Math.max(...t) * (a + b) * 2

  while (left <= right) {
    let mid = Math.floor((left + right) / 2)

    if (canBuild(a, b, g, s, w, t, mid)) right = mid - 1
    else left = mid + 1
  }

  return left
}

File: programmers/월간 코드 챌린지 시즌3/금과 은 운반하기/solution.py
import math


def solution(a, b, g, s, w, t):
    def can_build(time):
        S, G, total = 0, 0, 0

        for i in range(len(t)):
            W = min(g[i] + s[i], w[i] * math.floor(time / (2 * t[i]) + 0.5))

            S += min(s[i], W)
            G += min(g[i], W)
            total += W

        return a <= G and b <= S and a + b <= total

    left = 1
    right = max(t) * (a + b) * 2
    while left <= right:
        mid = math.floor((left + right) / 2)

        if can_build(mid):
            right = mid - 1
        else:
            left = mid + 1

    return left
